### PR TITLE
updated w/ final implementation and tests

### PR DESCRIPTION
**Pull request by brunochristensen**

### CODE REVIEW

**Review by will-eand — COMMENTED** [review]
General comment: I think your methods are both very concise and your variables names are chosen good. It was difficult to find glaring issues in your logic. There were a few edge cases not considered in your test file and I made comments that addressed those topics. Great job, especially on the part where you checked for duplicates by making a helper method using the set method.

File: src/main/java/StringManipulation.java [hunk 1]
```diff
@@ -1,33 +1,87 @@
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.Set;
+
 public class StringManipulation implements StringManipulationInterface {
 
+    String s;
```
Comment (will-eand) [inline src/main/java/StringManipulation.java:8]: On line 8, the data field String s should be private. The business logic of the code needs to incorporate encapsulation to protect from external access.

File: src/main/java/StringManipulation.java [hunk 2]
```diff
@@ -1,33 +1,87 @@
...
     @Override
     public String getString() {
-        return null;
```
Comment (will-eand) [inline src/main/java/StringManipulation.java:5]: This is methods breaks and the return method never gets accessed due to always returning null.

File: src/main/java/StringManipulation.java [hunk 3]
```diff
@@ -1,33 +1,87 @@
...
+        if (s == null || s.isEmpty())
+            return 0;
+        return s.split("\\s+").length;
     }
 
     @Override
     public String removeNthCharacter(int n, boolean maintainSpacing) {
-        return null;
+        if (n <= 0) {
+            throw new IllegalArgumentException();
```
Comment (will-eand) [inline src/main/java/StringManipulation.java:29]: It may be helpful to provide am message pertaining to the exact reason illegalArgumentException is called. This may help clarify during testing to save some time.

File: src/main/java/StringManipulation.java [hunk 4]
```diff
@@ -1,33 +1,87 @@
...
+            throw new IndexOutOfBoundsException();
+        }
+        return Arrays.copyOfRange(words, startWord - 1, endWord);
     }
 
     @Override
-    public String restoreString(int[] indices) {
-        return null;
+    public String restoreString(int[] indices){
+        if(indices.length != s.length() || duplicates(indices)){
```
Comment (will-eand) [inline src/main/java/StringManipulation.java:65]: Utilizing a set here to check for duplicates is nice way and optimal way to compare for duplicates. You helped me realize that I might not have done this for my own method. Well done.

File: src/test/java/StringManipulationTest.java [hunk 5]
```diff
@@ -118,32 +156,45 @@ public void testRestoreString1()
...
     @Test
     public void testRestoreString5()
     {
-        fail("Not yet implemented");
-
+        manipulatedstring.setString("art");
+        int [] array;
+        array=new int[]{1,0,2,3};
+        assertThrows(IllegalArgumentException.class, () -> manipulatedstring.restoreString(array));
     }
```
Comment (will-eand) [inline src/test/java/StringManipulationTest.java:199]: Your code handles illegalArgumentException in two situations; the first is when there are duplicates, and the second when the size of the array is either too big or too small relative to the length of the string. I see you testing the duplicates well, but what about different length of arrays being inputted? Another testRestoreString6 could be here.

File: src/test/java/StringManipulationTest.java [hunk 6]
```diff
@@ -53,59 +62,88 @@ public void testRemoveNthCharacter2() {
...
     @Test
-    public void testGeSubStrings3() {
-        fail("Not yet implemented");
+    public void testGetSubStrings3() {
+        manipulatedstring.setString("This is my string");
+        assertThrows(IllegalArgumentException.class, () -> manipulatedstring.getSubStrings(3, 2));
     }
+
+    //This test handles IllegalArgumentException for improper start and end values
     @Test
```
Comment (will-eand) [inline src/test/java/StringManipulationTest.java:129]: This test case is redundant. The test above tests the exact same thing as below. you're Testing the situation where the startWord is greater than the endWord, but there is also the possibility the start word and end words are the same, OR that the parameter inputs a negative number on accident. These are other cases you can test.

**Review by will-eand — APPROVED** [review]
